### PR TITLE
Added a lint rule to detect missing voice definitions

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/7136.

### CODE REVIEW

File: OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		{
+			foreach (var actorInfo in map.Rules.Actors)
+			{
+				var voiceSet = string.Empty;
+
+				foreach (var traitInfo in actorInfo.Value.Traits.WithInterface<ITraitInfo>())
+				{
+					var fields = traitInfo.GetType().GetFields();
+					foreach (var field in fields)
+					{
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Can drop these braces.

File: OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+						if (field.HasAttribute<VoiceSetReferenceAttribute>())
+							voiceSet = LintExts.GetFieldValues(traitInfo, field, emitError).FirstOrDefault();
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Should we give an error if an actor has more than one voice set defined?

File: OpenRA.Mods.Common/Traits/AttackMove.cs [hunk 3]
```diff
@@ -18,6 +18,8 @@ namespace OpenRA.Mods.Common.Traits
 	[Desc("Provides access to the attack-move command, which will make the actor automatically engage viable targets while moving to the destination.")]
 	class AttackMoveInfo : ITraitInfo
 	{
+		[VoiceReference] public readonly string AttackMoveVoice = "Move";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/AttackMove.cs]: All these prefixes are redundant, since they exist inside their own trait namespace.  I suggest uniformly using `Voice`.

File: OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs (hunk 2 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Hadn't thought about that. Maybe a warning?

**Comment by pchote** [comment]
Looks sensible overall, but lets converge on a consistent naming scheme for these new parameters before merging.

File: OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs (hunk 2 above)
Comment (penev92) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Voices can easily be switched on upgrade, along with artwork. The code doesn't care about any voice set other than the first one at the moment, but could easily be changed to handle that case.
So multiple voice sets on a single actor sounds perfectly plausible to me.

Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Good point @penev92.  We will need to do the checks below for each voice set on the actor, then.

File: OpenRA.Mods.Common/Traits/Voiced.cs [hunk 4]
```diff
@@ -19,7 +19,9 @@ namespace OpenRA.Mods.Common.Traits
 	public class VoicedInfo : ITraitInfo
 	{
 		[Desc("Which voice set to use.")]
-		[VoiceReference] public readonly string VoiceSet = null;
+		[VoiceSetReference] public readonly string VoiceSet = null;
+
+		[VoiceReference] public readonly string SelectVoice = "Select";
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: Maybe move this...

File: OpenRA.Mods.Common/Traits/Voiced.cs [hunk 5]
```diff
@@ -37,6 +39,7 @@ public Voiced(Actor self, VoicedInfo info)
 		}
 
 		public string VoiceSet { get { return Info.VoiceSet; } }
+		public string SelectVoice { get { return Info.SelectVoice; } }
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: ...and this to the Selectable trait instead? Then you won't need the prefix, either.

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: Nah. We just split the voices off. I'd personally rather keep the voices in `Voiced`, even this one.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: I agree with @reaperrr.  There's no need to make this a special case.

**Comment by Mailaender** [comment]
Updated with the suggestions made.

File: OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+			foreach (var actorInfo in map.Rules.Actors)
+			{
+				var voiceSet = string.Empty;
+
+				foreach (var traitInfo in actorInfo.Value.Traits.WithInterface<ITraitInfo>())
+				{
+					var fields = traitInfo.GetType().GetFields();
+					foreach (var field in fields)
+						if (field.HasAttribute<VoiceSetReferenceAttribute>())
+							voiceSet = LintExts.GetFieldValues(traitInfo, field, emitError).FirstOrDefault();
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: This still doesn't look right.  Any traits that are defined in the yaml before the `Voiced` trait will not be checked.

Comment (Mailaender) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: There is only the voiced trait so it does not matter.

Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Let me rephrase: The linter will not test any traits with `VoiceReference` fields that are defined on the actor above the `Voiced` trait definition.  The following will not be detected by the linter (try it yourself!):
```
AnyTraitThatDefinesVoices:
	Voice: Bogus
Voiced:
	VoiceSet: Whatever
```

Comment (Mailaender) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: VoiceSetReferenceAttribute != VoiceReferenceAttribute and I am looping through the fields twice here getting the VoiceSet first.

Comment (Mailaender) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Oh wait, I am only looping through the fields of the same trait.

Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Looping over the fields twice only helps if the same trait defines both `VoiceSetReferenceAttribute` and `VoiceReferenceAttribute`.  The problem occurs for *traits* that are parsed before the first VoiceSetReference.

Try it yourself and see: add `Voice: Bogus` to the TS Wolverine's `AttackFrontal` trait.  The linter won't detect it because it is above `Voiced`, and the game will crash as soon as you try to attack something.

Comment (Mailaender) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: That was stupid. Updated accordingly.

Comment (Mailaender) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: You commented the wrong line. That confused me and is also the reason this comment won't hide although I fixed it.

**Comment by Mailaender** [comment]
Had to rebase this due to MiniYaml file conflicts.

File: OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace OpenRA.Mods.Common.Lint
+{
+	public class CheckVoiceReferences : ILintPass
+	{
+		public void Run(Action<string> emitError, Action<string> emitWarning, Map map)
+		{
+			foreach (var actorInfo in map.Rules.Actors)
+			{
+				var voiceSet = string.Empty;
```
Comment (penev92) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: This here (actually the line that sets it below, but that is too overcrowded already) assumes the actor will have only one voice set, and will ignore all but the first if there are multiple. Shouldn't be too hard to remove the `.FirstOrDefault()` from [line 31](https://github.com/OpenRA/OpenRA/pull/8412/files#diff-53fde2db49577802d0410cae9f4d5c2dR31) and add a `foreach(var voiceSet in voiceSets)` around lines [44](https://github.com/OpenRA/OpenRA/pull/8412/files#diff-53fde2db49577802d0410cae9f4d5c2dR44) through 53 (after you rename this to `voiceSets`, of course).

**Comment by penev92** [comment]
All right! I like the cleanup of hardcoded strings very much.
Code looks good. I still haven't reviewed the YAML changes in detail.
Three things I noticed that need attention:
 - Fixing the upgrade rule.
 - Making the check for all voice sets on the actor.
 - Fixing `DeathSounds` in all 4 mods - you have an upgrade rule, but you didn't run it and the YAML still contains definitions like `DeathSound: Zapped`, which should now be `Voice: Zapped`.

Pinging all available reviewers, as this is a fairly big PR (don't let the numbers GitHub shows fool you, it's not *that* big) and conflicts with at least several other PRs, so would be better to get it merged ASAP.

File: OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs (hunk 7 above)
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: I second this request (actually this is exactly what I asked for in the first review pass, so you're seconding me :)).

**Comment by Mailaender** [comment]
Updated.

**Comment by Mailaender** [comment]
Rebased.

File: OpenRA.Game/Selection.cs [hunk 8]
```diff
@@ -59,7 +59,11 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
 
 			var voicedActor = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && a.HasTrait<IVoiced>());
 			if (voicedActor != null)
-				voicedActor.PlayVoice("Select");
+			{
+				var selectable = voicedActor.TraitOrDefault<Selectable>();
+				if (selectable != null)
```
Comment (pchote) [inline OpenRA.Game/Selection.cs]: This should be merged into the `voicedActor` query to ensure that the selected actor will always play a voice.

File: OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+		public void Run(Action<string> emitError, Action<string> emitWarning, Map map)
+		{
+			foreach (var actorInfo in map.Rules.Actors)
+			{
+				foreach (var traitInfo in actorInfo.Value.Traits.WithInterface<ITraitInfo>())
+				{
+					var fields = traitInfo.GetType().GetFields();
+					foreach (var field in fields)
+					{
+						if (field.HasAttribute<VoiceSetReferenceAttribute>())
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Style nit: You can avoid one block level here and in `CheckVoices` by adding `.Where(f => f.HasAttribute<VoiceSetReferenceAttribute>())` to the end of the `fields` definition.

File: OpenRA.Mods.Common/Lint/LintExts.cs [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+using System;
+using System.Reflection;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Lint
+{
+	public class LintExts
+	{
+		public static string[] GetFieldValues(ITraitInfo traitInfo, FieldInfo fieldInfo, Action<string> emitError)
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/LintExts.cs]: It doesn't look like you depend on anything `ITraitInfo` specific in here, so maybe change this to `object`?

File: mods/cnc/rules/infantry.yaml [hunk 11]
```diff
@@ -15,6 +15,7 @@ E1:
 	Armament:
 		Weapon: M16
 	AttackFrontal:
+		Voice: Move
```
Comment (pchote) [inline mods/cnc/rules/infantry.yaml]: It may be cleaner to define an Attack voice instead of overriding this everywhere.

Comment (pchote) [inline mods/cnc/rules/infantry.yaml]: Otherwise, `Move` should really be renamed to something more generic like `Action`.

File: OpenRA.Mods.Common/Lint/LintExts.cs (hunk 10 above)
Comment (penev92) [inline OpenRA.Mods.Common/Lint/LintExts.cs]: Yes, I actually copied this method for my upgrades lint check and had to change this to `object`.
So feel free to change or leave it - it's going to change anyway.

File: OpenRA.Mods.Common/Traits/Harvester.cs [hunk 12]
```diff
@@ -270,7 +281,7 @@ public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool qu
 
 		public string VoicePhraseForOrder(Actor self, Order order)
 		{
-			return (order.OrderString == "Harvest" || (order.OrderString == "Deliver" && !IsEmpty)) ? "Move" : null;
+			return (order.OrderString == "Harvest" || (order.OrderString == "Deliver" && !IsEmpty)) ? info.Voice : null;
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Harvester.cs]: Does this mean you'd have the same (say, "Harvest") definition played when you order the harvester to harvest a resource and when you order this harvester to return with said resource? RA2 already had both Harvest and Return separated.

Comment (penev92) [inline OpenRA.Mods.Common/Traits/Harvester.cs]: For the moment - yes. Looks like it won't be any problem separating them, but I'll leave that up to @Mailaender, as this PR isn't small or simple, and that's not really related.

File: mods/cnc/rules/infantry.yaml (hunk 11 above)
Comment (Mailaender) [inline mods/cnc/rules/infantry.yaml]: > It may be cleaner to define an Attack voice instead of overriding this everywhere.

That will add redundancy.

> Otherwise, `Move` should really be renamed to something more generic like `Action`.

I decided against it for compatibility reasons and pull request scope.

File: OpenRA.Mods.Common/Lint/LintExts.cs (hunk 10 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Lint/LintExts.cs]: Changed it to `object` already.

File: OpenRA.Mods.Common/Traits/Harvester.cs (hunk 12 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Harvester.cs]: I separated those while I was at it.

**Comment by Mailaender** [comment]
Updated as requested.

**Comment by GraionDilach** [comment]
TD mod still has DeathSound defined.

**Comment by Mailaender** [comment]
Fixed.

**Comment by GraionDilach** [comment]
Cannot find any other issues.

File: mods/cnc/rules/infantry.yaml (hunk 11 above)
Comment (pchote) [inline mods/cnc/rules/infantry.yaml]: You're arguing 21 that duplicated `Voice: Move` definitions is less redundant than vs 1 duplicated voice definition.

File: OpenRA.Game/Selection.cs [hunk 13]
```diff
@@ -57,9 +57,13 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
...
-			var voicedActor = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && a.HasTrait<IVoiced>());
-			if (voicedActor != null)
-				voicedActor.PlayVoice("Select");
+			var voicedActors = actors.Where(a => a.Owner == world.LocalPlayer && a.IsInWorld && a.HasTrait<IVoiced>());
+			var selectableActors = voicedActors.ToDictionary(a => a, v => v.TraitOrDefault<Selectable>());
+			foreach (var selectable in selectableActors.Where(v => v.Value != null))
+			{
+				selectable.Key.PlayVoice(selectable.Value.Info.Voice);
+				break;
+			}
```
Comment (pchote) [inline OpenRA.Game/Selection.cs]: This can be simplified to 
```csharp
// Play the selection voice from one of the selected actors
// TODO: This probably should only be considering the newly selected actors
// TODO: Ship this into an INotifySelection trait to remove the engine dependency on Selectable
foreach (var a in actors)
{
	if (a.Owner != world.LocalPlayer || !a.IsInWorld)
		continue;

	var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
	if (selectable == null || !a.HasVoice(selectable.Voice))
		continue;

	a.PlayVoice(selectable.Voice); 
	break;
}
```

Comment (penev92) [inline OpenRA.Game/Selection.cs]: +1 to that.
But change `a in actors` to `a in newSelection` and remove the first `TODO`

File: OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+				var fields = traitInfo.GetType().GetFields().Where(f => f.HasAttribute<VoiceReferenceAttribute>());
+				foreach (var field in fields)
+				{
+					var voices = LintExts.GetFieldValues(traitInfo, field, emitError);
+					foreach (var voice in voices)
+					{
+						if (string.IsNullOrEmpty(voice))
+							continue;
+
+						var soundInfo = map.Rules.Voices[voiceSet.ToLowerInvariant()];
```
Comment (penev92) [inline OpenRA.Mods.Common/Lint/CheckVoiceReferences.cs]: Minor nit: This can easily go to the top of the method.

**Comment by penev92** [comment]
I like the code changes and cleanup.
But after reading the entire diff I have to agree with @pchote's earlier comment about the `Attack` voice - would look better to have such a voice in TD and TS (and bring it back for D2k and RA). A lot of YAML duplication aside, it defeats the purpose of having a default value if **everything** is going to override it.
Lint check works, and I can still hear stuff ingame (wow, what a surprise :D ), so :+1: with the above comments.

**Comment by pchote** [comment]
If none of the mods use separate attack and move voices for the generic units (heros are of course an exception) then my earlier suggestion of changing the defaults to "Action" and removing all the overrides seems like the best approach.

**Comment by penev92** [comment]
Needs a rebase now that #8388 was merged. Sorry.

**Comment by GraionDilach** [comment]
TS uses separate voices for Attack and Move. Notsure on D2k.

**Comment by Mailaender** [comment]
Rebased and special cased Tiberian Sun instead of Dune 2000, Red Alert, Tiberian Dawn. :sweat:

**Comment by reaperrr** [comment]
Looks good to me. I didn't test it thoroughly, but the move/attack differentiation in TS worked, and I didn't hear any obvious regressions in TD, D2k and RA either.

:+0.9: from me.

**Comment by pchote** [comment]
I haven't tested this, but this looks like a really nice improvement over what we currently have.

**Comment by Mailaender** [comment]
In summary: it removes magic behavior, adds automatic checks so all the rule refactors that almost always introduced hard to spot regressions will get tested. I would like to see this in before we do more large scale MiniYaml changes.

**Comment by obrakmann** [comment]
Looks ok to me, :+1: 

A future PR could also add checks to see if the VoiceSet exists at all (currently there wouldn't be any message at all).

**Comment by Mailaender** [comment]
I think we already do that: https://github.com/OpenRA/OpenRA/pull/8412/files#diff-2908d570caac7d047ea2dd3ea940cb28R40

**Comment by obrakmann** [comment]
Whoops, missed that. Even better then.

[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/fbdc527380279b68739b43a61be3a2c7b4eb956b%5E...fbdc527380279b68739b43a61be3a2c7b4eb956b)
